### ISSUE TITLE
TrackParticle Data Units

### ISSUE THREAD

**Issue by arsampath** [SmileiPIC/Smilei#120]
Dear developers,

We ran a simple 2D test (Smilei-4.1) with a gaussian spatial density distribution in the longitudinal direction for an electron beam. For this, we use the TrackParticle Diagnostics to deposit 'x' and 'w' (weight) at every time step. 
For this distribution, we want to calculate the standard deviation. We tried to do it as follows: 
1. Weighted mean, x_w = sum(w[i]*x[i])/sum(wt[i])
2. Standard deviation = sum( w[i]* (x[i] - x_w)^2) / ( sum(w[i]) * ( len(w) -1.) ) / len(w)
 We are able to reproduce the mean that we specify in our density profile. 
However standard deviation is not as we expect. Is this because of normalisation or how weights are defined? Or is it that we are calculating it incorrectly? Please explain this to us.

Thank you for your time.

**Comment by mccoys** [comment]
In which sense is it not what you expected ?
Have you compared the profile of density to the requested profile ?

Note that TrackParticles are not the best way to make weighted averages. You may want to look at ParticleBinning to compute distribution and averages. It can save you some disk space.

**Comment by arsampath** [comment]
Thank you for the response. 

I am attaching the input file which I used to make this small test. 
We have used the inbuilt gaussian profile by SMILEI. 
RMS_width (both in x and y) = 2.23 um.
The corresponding FWHM was calculated as 2*sqrt(2 ln2)* RMS_width and inputted in SMILEI profile. 
We then read data and process as follows: 
 1. track = S.TrackParticles("electron_beam").get()
 2. wt = np.array(track['w'][0])
 3. x = np.array(track['x'][0])
 4. x_wt = x*wt
 5. s = np.sum(x_wt)
 6. mean =  s/np.sum(wt) 
 7. result =0.
 8. for index, i in enumerate(x):
	       result+=  wt[index]*(i-mean)**2 
 9. result = result / ((len(wt)-1.) * np.sum(wt)) / len(wt)
10. print 'res ', np.sqrt(result) *  S.namelist.l_skin 

This final value should be 2.23um (our input), the rms width. But this does not match for us. 

[test_input.txt](https://github.com/SmileiPIC/Smilei/files/3260231/test_input.txt)

**Comment by mccoys** [comment]
This is probably because you don't have the full gaussian distribution in the box so you cannot compute the correct rms. If you check your profile shape from your ParticleBinning(0), you will find the correct distribution.

**Comment by arsampath** [comment]
We tried to extract the data from Particle Binning diagnostics. 

P_bin = S.ParticleBinning(0).get(0)
x = np.array(P_bin['x'])
wt_bin = np.array(P_bin['data'][0])
mean = np.sum(x*wt_bin) / np.sum(wt_bin)
count = 0.
for i in wt_bin:
         if(i!=0.): count+=1.
result = 0.
for ind, i in enumerate(x):
        result+=wt_bin[ind]*(i-mean)**2.
result = result / ( (count -1.) * np.sum(wt_bin) ) / np.sum(wt_bin)
print sqrt(result) * S.namelist.l_skin

This diagnostic result also does not give expected rms width of the same order (which is um). 

Kindly let us know how to proceed.

**Comment by mccoys** [comment]
Again, do not use RMS because your profile does not fit in the box.

Plot the profile:
```
P=S.ParticleBinning(0)
x=P.getAxis('x')
profile=vectorize(S.namelist.Species[0].number_density)
y=profile(x,S.namelist.b_ycenter)
P.plot()
plot(x,y/y.max()*P.getData()[0].max())
```

**Comment by arsampath** [comment]
We would like to know how we can use Smilei's post-processing tools to get some derived quantities. 
For example, if we want to calculate beam emittance given by the following formula:
![image](https://user-images.githubusercontent.com/27001896/59061037-bf797b80-88a2-11e9-84ee-46e1c770832d.png) 
Then we need to know the RMS value of transverse spatial coordinates, the RMS value of the ratio of transverse momentum to the longitudinal momentum and the co-relation term. 

In short, we would like to know how to use SMILEI's post-processing tool to get any derived quantities (quantities we can calculate from spatial and momentum coordinates).
Thank you for your time.

**Comment by mccoys** [comment]
Let me say it again: your profile does not fit in the box, which is the reason why your rms value is incorrect. As a consequence, the problem is not the post-processing (you do it correctly), but that your simulation is not adequate. You need a bigger simulation box, or a smaller beam.

Concerning your emittance formula, please specify what is `x'`.

**Comment by Mi3Sangal** [comment]
[input.txt](https://github.com/SmileiPIC/Smilei/files/3266246/input.txt)
![Distribution](https://user-images.githubusercontent.com/34055763/59110061-8d1d5c00-893e-11e9-84e5-ec9df70b81b1.png)

Dear Developers,

I have also been working on the same thing and have been following this thread.  
Thank you for your help with the calculation of rms. You are right and it seems to work properly if the simulation box is large enough. 

However, here I have attached and input file for just initializing a beam with a Gaussian spatial density profile. 

In my input file. The reference number density = n0_target_value. (approximately 1.e29 m^-3).
n0_beam_normalised = 1.e26/n0_target_value = 5.e-4 (approx). 
 (Here 1.e26 is the electron no. density of the beam in units of m^-3).

I also specify the density profile to be Gaussian using Smilei's built-in profile.

Thus, when I plot the lineout of the charge density "Rho_electron_beam" using the Fields diagnostic (the figure is attached), I can see that the peak value of this distribution is lower than what is specified in the input. This is approximately 3.e-4. 

Could you please let me know what could be going wrong? 

Besides this, if I use the ParticleBinning Diagnostic to plot the spatial distribution, the peak value is different. This I understand is because how "weight" is defined for Particle Binning. 
According to my understanding, this already accounts for the cell hypervolume. Hence in the plot : 

Weight = number density (in units of Nr)/No. of macro particles per cell. Is this correct?
However, this also does not give me the expected peak value of no. density of the beam. 

Could this be a problem with my post-processing or with the initialization in the code?
Thank you for your help.

**Comment by mccoys** [comment]
The issue in your situation is that you use `yorder=0`. A gaussian with order 0 makes little sense, and the result is something like `exp(-x**0 / s)`, resulting in a factor close to 2.

You should put `yfwhm=1000000.` instead.

**Comment by arsampath** [comment]
Thank you for your help. The density profile is correct now.
